pkg/telemetry/stargate_prometheus_resourcer.go

```diff
@@ -0,0 +1,252 @@
+	return &sm, nil
+}
+
+// UpdateResources executes the creation of the desired Prometheus resources on the cluster.
+func (cfg StargatePrometheusResourcer) UpdateResources(ctx context.Context, client runtimeclient.Client, owner *stargateapi.Stargate) error {
+	desiredSM, err := cfg.NewServiceMonitor()
+	if err != nil {
+		return err
+	}
+	cfg.StargateTelemetryResourcer.Logger.Info("checking whether Prometheus ServiceMonitor for Stargate already exists")
+	// Logic to handle case where SM does not exist.
+	actualSM := &promapi.ServiceMonitor{}
+	if err := client.Get(ctx, types.NamespacedName{Name: desiredSM.Name, Namespace: desiredSM.Namespace}, actualSM); err != nil {
+		if errors.IsNotFound(err) {
+			cfg.StargateTelemetryResourcer.Logger.Info("Prometheus ServiceMonitor for Cassandra not found, creating")
+			if err := controllerutil.SetControllerReference(owner, desiredSM, client.Scheme()); err != nil {
+				cfg.StargateTelemetryResourcer.Logger.Error(err, "could not set controller reference for ServiceMonitor", "owner", owner)
+				return err
+			} else if err = client.Create(ctx, desiredSM); err != nil {
+				if errors.IsAlreadyExists(err) {
+					// the read from the local cache didn't catch that the resource was created already; simply requeue until the cache is up-to-date
+					return nil
+				} else {
+					cfg.StargateTelemetryResourcer.Logger.Error(err, "could not create ServiceMonitor resource", "resource", desiredSM, "owner", owner)
+					return err
+				}
+			}
+			return nil
+		} else {
+			cfg.StargateTelemetryResourcer.Logger.Error(err, "could not get ServiceMonitor resource")
+			return err
+		}
+	}
+	// Logic to handle case where SM exists, but is in the wrong state.
+	actualSM = actualSM.DeepCopy()
+	if !annotations.CompareAnnotations(actualSM, desiredSM, k8ssandraapi.ResourceHashAnnotation) {
+		resourceVersion := actualSM.GetResourceVersion()
+		desiredSM.DeepCopyInto(actualSM)
+		actualSM.SetResourceVersion(resourceVersion)
+		if err := controllerutil.SetControllerReference(owner, actualSM, client.Scheme()); err != nil {
+			cfg.StargateTelemetryResourcer.Logger.Error(err, "could not set controller reference for ServiceMonitor", "resource", desiredSM, "owner", owner)
+			return err
+		} else if err := client.Update(ctx, actualSM); err != nil {
+			cfg.StargateTelemetryResourcer.Logger.Error(err, "could not update ServiceMonitor resource", "resource", desiredSM, "owner", owner)
+			return err
+		} else {
+			cfg.StargateTelemetryResourcer.Logger.Info("successfully updated the Stargate ServiceMonitor")
+			return nil
+		}
+	}
+	return nil
+}
+
+// CleanupResources executes the cleanup of any resources on the cluster, once they are no longer required.
+// CleanupResources does not rely on user-defined CommonLabels to select which resources to delete. It uses only
+// k8ssandra-operator defined labels.
+func (cfg StargatePrometheusResourcer) CleanupResources(ctx context.Context, client runtimeclient.Client) error {
```

Comment:
This function is virtually the same as `CassPrometheusResourcer.CleanupResources`. With a bit of refactoring a lot of duplicate code can be eliminated.

Reply:
Done.